server/channels/store/sqlstore/channel_store.go

```diff
@@ -1922,6 +1922,82 @@ func (s SqlChannelStore) UpdateMemberNotifyProps(channelID, userID string, props
 	return dbMember.ToModel(), err
 }
 
+// PatchMultipleMembersNotifyProps updates the NotifyProps of multiple channel members at once without modifying
+// unspecified fields. Like UpdateMemberNotifyProps, it doesn't modify the LastUpdateAt of affected rows.
+//
+// Note that the returned array may not be in the same order as the provided IDs.
+func (s SqlChannelStore) PatchMultipleMembersNotifyProps(members []*model.ChannelMemberIdentifier, notifyProps map[string]string) ([]*model.ChannelMember, error) {
+	if len(notifyProps) == 0 {
+		return nil, errors.New("PatchMultipleMembersNotifyProps: No notifyProps specified")
+	}
+
+	if err := model.IsChannelMemberNotifyPropsValid(notifyProps, true); err != nil {
+		return nil, err
+	}
+
+	transaction, err := s.GetMasterX().Beginx()
+	if err != nil {
+		return nil, errors.Wrap(err, "begin_transaction")
+	}
+	defer finalizeTransactionX(transaction, &err)
+
+	// Make the where clause first since it'll be used multiple times
+	whereClause := sq.Or{}
```

Comment:
I'd probably expect a `IN` operator to be more efficient here but may benefit some benchmarking on a load-test dataset to confirm.

Reply:
According to the maintainers of Squirrel (https://github.com/Masterminds/squirrel#faq), this should produce the same query plan as `IN` since we're checking multiple fields at once.

If we want to benchmark it, I don't know what our plans are for benchmarking the MS Teams plugin (this is specifically going to be used to let users automute linked channels), but we could look at benchmarking this at the same time

Reply:
I mean purely the raw query, we can construct it manually and run it against one of our load-test dumps and verify the plans are equally efficient. Nothing more fancy than that and @agnivade should be able to help.

Reply:
@hmhealey - If you can just give me a sample raw query, I can test out the OR version and IN version to confirm that the query plans are same.

Reply:
Based on a run of the store unit tests, the two queries run on Postgres are:
```sql
UPDATE
    ChannelMembers
SET
    notifyprops = notifyprops || '{"channel_auto_follow_threads":"off","test_key":"test_value"}'::jsonb,
    LastUpdateAt = '1704832400298'
WHERE
    ((ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw') OR
     (ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'i9iasntzit8d7y5cpok5wk1bco') OR
     (ChannelId = 'dgfx5bzyq7nrukerwdcjio93mc' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw'));

SELECT
    ChannelMembers.ChannelId,
    ChannelMembers.UserId,
    ChannelMembers.Roles,
    ChannelMembers.LastViewedAt,
    ChannelMembers.MsgCount,
    ChannelMembers.MentionCount,
    ChannelMembers.MentionCountRoot,
    COALESCE(ChannelMembers.UrgentMentionCount, 0) AS UrgentMentionCount,
    ChannelMembers.MsgCountRoot,
    ChannelMembers.NotifyProps,
    ChannelMembers.LastUpdateAt,
    ChannelMembers.SchemeUser,
    ChannelMembers.SchemeAdmin,
    ChannelMembers.SchemeGuest,
    TeamScheme.DefaultChannelGuestRole TeamSchemeDefaultGuestRole,
    TeamScheme.DefaultChannelUserRole TeamSchemeDefaultUserRole
    TeamScheme.DefaultChannelAdminRole TeamSchemeDefaultAdminRole,
    ChannelScheme.DefaultChannelGuestRole ChannelSchemeDefaultGuestRole,
    ChannelScheme.DefaultChannelUserRole ChannelSchemeDefaultUserRole,
    ChannelScheme.DefaultChannelAdminRole ChannelSchemeDefaultAdminRole
FROM
    ChannelMembers
    INNER JOIN Channels ON ChannelMembers.ChannelId = Channels.Id
    LEFT JOIN Schemes ChannelScheme ON Channels.SchemeId = ChannelScheme.Id
    LEFT JOIN Teams ON Channels.TeamId = Teams.Id
    LEFT JOIN Schemes TeamScheme ON Teams.SchemeId = TeamScheme.Id
WHERE
    ((ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw') OR
     (ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'i9iasntzit8d7y5cpok5wk1bco') OR
     (ChannelId = 'dgfx5bzyq7nrukerwdcjio93mc' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw'));
```
and on MySQL, they're
```sql
SELECT
    ChannelMembers.ChannelId,
    ChannelMembers.UserId,
    ChannelMembers.Roles,
    ChannelMembers.LastViewedAt,
    ChannelMembers.MsgCount,
    ChannelMembers.MentionCount,
    ChannelMembers.MentionCountRoot,
    COALESCE(ChannelMembers.UrgentMentionCount, 0) AS UrgentMentionCount,
    ChannelMembers.MsgCountRoot,
    ChannelMembers.NotifyProps,
    ChannelMembers.LastUpdateAt,
    ChannelMembers.SchemeUser,
    ChannelMembers.SchemeAdmin,
    ChannelMembers.SchemeGuest,
    TeamScheme.DefaultChannelGuestRole TeamSchemeDefaultGuestRole,
    TeamScheme.DefaultChannelUserRole TeamSchemeDefaultUserRole
    TeamScheme.DefaultChannelAdminRole TeamSchemeDefaultAdminRole,
    ChannelScheme.DefaultChannelGuestRole ChannelSchemeDefaultGuestRole,
    ChannelScheme.DefaultChannelUserRole ChannelSchemeDefaultUserRole,
    ChannelScheme.DefaultChannelAdminRole ChannelSchemeDefaultAdminRole
FROM
    ChannelMembers
    INNER JOIN Channels ON ChannelMembers.ChannelId = Channels.Id
    LEFT JOIN Schemes ChannelScheme ON Channels.SchemeId = ChannelScheme.Id
    LEFT JOIN Teams ON Channels.TeamId = Teams.Id
    LEFT JOIN Schemes TeamScheme ON Teams.SchemeId = TeamScheme.Id
WHERE
    ((ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw') OR
     (ChannelId = 'crtoo6rex7gs5nowwkyyhhguwh' AND UserId = 'i9iasntzit8d7y5cpok5wk1bco') OR
     (ChannelId = 'dgfx5bzyq7nrukerwdcjio93mc' AND UserId = 'w14uyi84w7r3in8ifgjqc4eetw'));

UPDATE
    ChannelMembers
SET
    NotifyProps = JSON_SET(NotifyProps, '$.channel_auto_follow_threads', 'off', '$.test_key', 'test_value'),
    LastUpdateAt = '1704832400235'
WHERE
    ((ChannelId = 'cbx5ar3h6jyjjjz9pywbgiakew' AND UserId = 'f8bxtzjc5ir1zmyqgxtfpi4qnw') OR
     (ChannelId = 'cbx5ar3h6jyjjjz9pywbgiakew' AND UserId = 'ha33p3crn3no8etntdshp3bsro') OR
     (ChannelId = 'jo9pfyei47gkdqpsph5bubj6jc' AND UserId = 'f8bxtzjc5ir1zmyqgxtfpi4qnw'));

SELECT
    ChannelMembers.ChannelId,
    ChannelMembers.UserId,
    ChannelMembers.Roles,
    ChannelMembers.LastViewedAt,
    ChannelMembers.MsgCount,
    ChannelMembers.MentionCount,
    ChannelMembers.MentionCountRoot,
    COALESCE(ChannelMembers.UrgentMentionCount, 0) AS UrgentMentionCount,
    ChannelMembers.MsgCountRoot,
    ChannelMembers.NotifyProps,
    ChannelMembers.LastUpdateAt,
    ChannelMembers.SchemeUser,
    ChannelMembers.SchemeAdmin,
    ChannelMembers.SchemeGuest,
    TeamScheme.DefaultChannelGuestRole TeamSchemeDefaultGuestRole,
    TeamScheme.DefaultChannelUserRole TeamSchemeDefaultUserRole,
    TeamScheme.DefaultChannelAdminRole TeamSchemeDefaultAdminRole,
    ChannelScheme.DefaultChannelGuestRole ChannelSchemeDefaultGuestRole,
    ChannelScheme.DefaultChannelUserRole ChannelSchemeDefaultUserRole,
    ChannelScheme.DefaultChannelAdminRole ChannelSchemeDefaultAdminRole
FROM
    ChannelMembers
    INNER JOIN Channels ON ChannelMembers.ChannelId = Channels.Id
    LEFT JOIN Schemes ChannelScheme ON Channels.SchemeId = ChannelScheme.Id
    LEFT JOIN Teams ON Channels.TeamId = Teams.Id
    LEFT JOIN Schemes TeamScheme ON Teams.SchemeId = TeamScheme.Id
WHERE
    ((ChannelId = 'cbx5ar3h6jyjjjz9pywbgiakew' AND UserId = 'f8bxtzjc5ir1zmyqgxtfpi4qnw') OR
     (ChannelId = 'cbx5ar3h6jyjjjz9pywbgiakew' AND UserId = 'ha33p3crn3no8etntdshp3bsro') OR
     (ChannelId = 'jo9pfyei47gkdqpsph5bubj6jc' AND UserId = 'f8bxtzjc5ir1zmyqgxtfpi4qnw'));
```

Reply:
All good :+1: Query plans are identical. Thanks Harrison!